arangod/Cluster/v8-cluster.cpp

```diff
@@ -1563,75 +1556,69 @@ static void JS_AsyncRequest(v8::FunctionCallbackInfo<v8::Value> const& args) {
   // Possible options:
   //   - coordTransactionID   (number)
   //   - timeout              (number)
-  //   - singleRequest        (boolean) default is false
   //   - initTimeout          (number)
 
-  auto cc = ClusterComm::instance();
+  TRI_GET_GLOBALS();
+
+  auto* pool = v8g->_server.getFeature<arangodb::NetworkFeature>().pool();
 
-  if (cc == nullptr) {
+  if (pool == nullptr) {
     TRI_V8_THROW_EXCEPTION_MESSAGE(
         TRI_ERROR_SHUTTING_DOWN,
-        "clustercomm object not found (JS_AsyncRequest)");
+        "connectionpool object not found (JS_AsyncRequest)");
   }
 
-  arangodb::rest::RequestType reqType;
+  fuerte::RestVerb reqType;
   std::string destination;
   std::string path;
-  auto body = std::make_shared<std::string>();
+  VPackBufferUInt8 body;
   std::unordered_map<std::string, std::string> headerFields;
   CoordTransactionID coordTransactionID = 0;
   double timeout = 0.0;
   double initTimeout = -1.0;
-  bool singleRequest = false;
-
-  PrepareClusterCommRequest(args, reqType, destination, path, *body, headerFields,
-                            coordTransactionID, timeout, singleRequest, initTimeout);
 
-  OperationID opId = cc->asyncRequest(coordTransactionID, destination, reqType,
-                                      path, body, headerFields, nullptr,
-                                      timeout, singleRequest, initTimeout);
-  ClusterCommResult res = cc->enquire(opId);
-  if (res.status == CL_COMM_BACKEND_UNAVAILABLE) {
-    TRI_V8_THROW_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL,
-                                   "couldn't queue async request");
-  }
+  PrepareClusterCommRequest(args, reqType, destination, path, body, headerFields,
+                            coordTransactionID, timeout, initTimeout);
+  
+  network::RequestOptions reqOpts;
+  reqOpts.retryNotFound = false;
+  reqOpts.timeout = network::Timeout(timeout);
+  reqOpts.skipScheduler = true;
+  
+  OperationID opId = TRI_NewTickServer();
+  auto ar = std::make_shared<::AsyncRequest>();
+  ar->destination = destination;
+  ar->operationID = opId;
+  ar->coordTransactionID = coordTransactionID;
+  
+  network::sendRequest(pool, destination, reqType, path, std::move(body), reqOpts)
+  .thenValue([ar] (network::Response&& r) {
+    {
+      std::lock_guard<std::mutex> guard(::_requestMutex);
+      ar->response = std::move(r);
+      ar->done = true;
+    }
+    ::_requestCV.notify_all();
```

Comment:
if we don't acquire the `_requestMutex` here, the notify_all() may be missed by the `wait_for` call in JS_Wait.

Reply:
nope

Reply:
why not?

Reply:
@jsteemann Pulled this from https://en.cppreference.com/w/cpp/thread/condition_variable/notify_all: "The notifying thread does not need to hold the lock on the same mutex as the one held by the waiting thread(s); in fact doing so is a pessimization, since the notified thread would immediately block again, waiting for the notifying thread to release the lock."

The basic idea is that if you are using the `condition_variable` to notify about a change to a variable state, and you protect access to the variable state with the `mutex`, and hold the `mutex` to wait for the `condition_variable`, then it should be okay. Essentially, the `mutex` guards against the case where you check the variable, then while preparing to wait, the other thread sets the variable and notifies before you wait.

Reply:
cause we synchronize the changes to the `::_requests` global var with the mutex